File: core/spec/browser-models.ts
/**
 * Contract for tab, session, and settings fields.
 * Harmony owns the runtime types in `entry/src/main/ets/models/BrowserModels.ets`.
 * Other clients must persist the same JSON keys. Adding a settings field still
 * requires the Harmony four-file set (interface, default, clone/normalize, decoder)
 * plus this spec.
 */

export type AppearanceMode = 0 | 1 | 2;
export type WebDarkModePreference = 0 | 1 | 2;
export type SitePermissionDecision = 0 | 1 | 2;
export type RuleStrength = 0 | 1;
export type SearchEngine = 0 | 1 | 2 | 3;
export type ReaderPaper = 0 | 1 | 2;
export type StartPageBackgroundSource = 0 | 1 | 2;
export type SuggestionKind = 0 | 1 | 2;
export type LoadErrorKind = 0 | 1 | 2 | 3 | 4 | 5;
export type SecurityState = 0 | 1 | 2 | 3;
export type UserAgentPreference = 0 | 1 | 2;
export type TabExpiry = 0 | 1 | 3 | 7;

export interface BlockStats {
  ads: number;
  trackers: number;
  malicious: number;
  popups: number;
  cookieBanners: number;
}

export interface CertificateSummary {
  subject: string;
  issuer: string;
  validFrom: string;
  validTo: string;
}

export interface BrowserTab {
  id: string;
  url: string;
  title: string;
  isPrivate: boolean;
  isLoading: boolean;
  progress: number;
  isDesktop: boolean;
  isPinned: boolean;
  isReader: boolean;
  canGoBack: boolean;
  canGoForward: boolean;
  blocked: BlockStats;
  faviconUrl: string;
  lastVisitedAt: number;
  scrollY: number;
  readerScrollY: number;
  scrollSavedAt: number;
  readerScrollSavedAt: number;
  formDraft: string;
  loadError: LoadErrorKind;
  securityState: SecurityState;
  certificate?: CertificateSummary;
}

export interface BrowserSession {
  tabs: BrowserTab[];
  activeTabId: string;
}

export interface SavedItem {
  id: string;
  title: string;
  url: string;
  createdAt: number;
  updatedAt: number;
  isRead: boolean;
  tags: string[];
}

export interface HistoryEntry {
  id: string;
  title: string;
  url: string;
  visitedAt: number;
  visitCount: number;
}

export interface SiteZoomRatio {
  host: string;
  percent: number;
}

export interface SiteUserAgentPreference {
  host: string;
  preference: UserAgentPreference;
}

export interface SitePermission {
  origin: string;
  camera: SitePermissionDecision;
  microphone: SitePermissionDecision;
  location: SitePermissionDecision;
  notifications: SitePermissionDecision;
}

export interface BrowserSettings {
  blockAds: boolean;
  blockTrackers: boolean;
  ruleStrength: RuleStrength;
  searchEngine: SearchEngine;
  customSearchTemplate: string;
  searchSuggestionsEnabled: boolean;
  gesturesEnabled: boolean;
  gestureActionsEnabled: boolean;
  gestureTabSwitchEnabled: boolean;
  gestureBlockingEnabled: boolean;
  autoHideToolbarEnabled: boolean;
  quickSitesEnabled: boolean;
  quickSiteLimit: number;
  startPageBackgroundEnabled: boolean;
  startPageBackgroundSource: StartPageBackgroundSource;
  customStartPageBackgroundUri: string;
  startPagePresetPortraitId: string;
  startPagePresetLandscapeId: string;
  tabExpiryDays: TabExpiry;
  historyRetentionDays: number;
  liveWebViewLimit: number;
  downloadConcurrency: number;
  largeDownloadThresholdMb: number;
  wifiOnlyDownloads: boolean;
  tabSoftLimit: number;
  minimumFontSize: number;
  siteZoomRatios: SiteZoomRatio[];
  defaultUserAgentPreference: UserAgentPreference;
  siteUserAgentPreferences: SiteUserAgentPreference[];
  privacyConsentAccepted: boolean;
  onboardingCompleted: boolean;
  appearance: AppearanceMode;
  webDarkMode: WebDarkModePreference;
  webDarkModeExcludedHosts: string[];
  sitePermissions: SitePermission[];
  clearCookiesOnTabClose: boolean;
  telemetryEnabled: boolean;
  readerFontSize: number;
  readerLineHeight: number;
  readerPaper: ReaderPaper;
}

export const HOME_URL = 'browser://home';
export const MAX_TAB_COUNT = 100;
export const LIVE_WEBVIEW_LIMIT_DEFAULT = 4;

File: ios/ZhuoBrowser/BrowserModels.swift
import Foundation

struct BrowserTab: Identifiable, Equatable, Codable {
    var id: String
    var url: String
    var title: String
    var isPrivate: Bool
    var isLoading: Bool
    var progress: Double
    var canGoBack: Bool
    var canGoForward: Bool
    var lastVisitedAt: TimeInterval
    var isReader: Bool
    var isDesktop: Bool
    var isPinned: Bool

    static func home(isPrivate: Bool) -> BrowserTab {
        BrowserTab(
            id: UUID().uuidString,
            url: URLPolicy.homeURL,
            title: isPrivate ? "无痕" : "新标签页",
            isPrivate: isPrivate,
            isLoading: false,
            progress: 0,
            canGoBack: false,
            canGoForward: false,
            lastVisitedAt: Date().timeIntervalSince1970,
            isReader: false,
            isDesktop: false,
            isPinned: false
        )
    }

    var displayTitle: String {
        if !title.isEmpty && title != url {
            return title
        }
        let host = URLPolicy.displayHost(url)
        return host.isEmpty ? (isPrivate ? "无痕" : "新标签页") : host
    }

    init(
        id: String,
        url: String,
        title: String,
        isPrivate: Bool,
        isLoading: Bool,
        progress: Double,
        canGoBack: Bool,
        canGoForward: Bool = false,
        lastVisitedAt: TimeInterval,
        isReader: Bool,
        isDesktop: Bool,
        isPinned: Bool = false
    ) {
        self.id = id
        self.url = url
        self.title = title
        self.isPrivate = isPrivate
        self.isLoading = isLoading
        self.progress = progress
        self.canGoBack = canGoBack
        self.canGoForward = canGoForward
        self.lastVisitedAt = lastVisitedAt
        self.isReader = isReader
        self.isDesktop = isDesktop
        self.isPinned = isPinned
    }

    enum CodingKeys: String, CodingKey {
        case id, url, title, isPrivate, isLoading, progress, canGoBack, canGoForward
        case lastVisitedAt, isReader, isDesktop, isPinned
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        url = try container.decode(String.self, forKey: .url)
        title = try container.decode(String.self, forKey: .title)
        isPrivate = try container.decode(Bool.self, forKey: .isPrivate)
        isLoading = try container.decodeIfPresent(Bool.self, forKey: .isLoading) ?? false
        progress = try container.decodeIfPresent(Double.self, forKey: .progress) ?? 0
        canGoBack = try container.decodeIfPresent(Bool.self, forKey: .canGoBack) ?? false
        canGoForward = try container.decodeIfPresent(Bool.self, forKey: .canGoForward) ?? false
        lastVisitedAt = try container.decode(TimeInterval.self, forKey: .lastVisitedAt)
        isReader = try container.decodeIfPresent(Bool.self, forKey: .isReader) ?? false
        isDesktop = try container.decodeIfPresent(Bool.self, forKey: .isDesktop) ?? false
        isPinned = try container.decodeIfPresent(Bool.self, forKey: .isPinned) ?? false
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(url, forKey: .url)
        try container.encode(title, forKey: .title)
        try container.encode(isPrivate, forKey: .isPrivate)
        try container.encode(isLoading, forKey: .isLoading)
        try container.encode(progress, forKey: .progress)
        try container.encode(canGoBack, forKey: .canGoBack)
        try container.encode(canGoForward, forKey: .canGoForward)
        try container.encode(lastVisitedAt, forKey: .lastVisitedAt)
        try container.encode(isReader, forKey: .isReader)
        try container.encode(isDesktop, forKey: .isDesktop)
        try container.encode(isPinned, forKey: .isPinned)
    }
}

struct PersistedSession: Codable {
    var tabs: [BrowserTab]
    var activeTabID: String
}
